ONSdigital/dp-api-clients-go
# Implement Cantabular GetDimensions and GetDimensionOptions with graphql

### What

- Implemented GetDimensions for Cantabular using graphQL (query all variables for a provided dataset)
- Implemented GetDimensionsByName for Cantabular using graphQL (query only provided variables for a dataset)
- Implemented GetDimensionOptions for Cantabular using graphQL (query categories for provided variables and dataset)
- Added unit tests

trello card: https://trello.com/c/P1WrnEct/5439-add-method-to-dp-api-clients-go-for-cantabular-flexible-datasets

used by:
- https://github.com/ONSdigital/dp-import-cantabular-dataset/pull/37
- https://github.com/ONSdigital/dp-import-cantabular-dimension-options/pull/36

### How to review

- Make sure code changes make sense
- Make sure unit tests pass

- (tried locally, no need to reproduce) You can run Cantabular server and API ext locally, and replace the unit test mocks with a real Cantabular connection like so, to validate that the client will behave as expected:
```
cantabularClient := cantabular.NewClient(
  cantabular.Config{
    Host:       "http://localhost:8491",
    ExtApiHost: "http://localhost:8492",
  },
  dphttp.NewClient(),
  nil,
)
```

__
(tried against large dataset in develop, no need to reproduce)
- dp ssh to cantabular api-ext: `dp ssh develop publishing 5 -p 8492:10.201.4.101:12100`
- Modified the `GetDimensionOptionsHappy` test to:
  - use a real client as described above
  - do the following request:
```
req := cantabular.StaticDatasetQueryRequest{
  Dataset:   "People-Households",
  Variables: []string{"oa"},
}

var m0, m1 runtime.MemStats

t0 := time.Now()
runtime.ReadMemStats(&m0)
resp, err := cantabularClient.GetDimensionOptions(testCtx, req)
runtime.ReadMemStats(&m1)
t1 := time.Since(t0)

fmt.Println("Time:      ", t1.String())
fmt.Println("Alloc:     ", m1.Alloc-m0.Alloc)
fmt.Println("TotalAlloc:", m1.TotalAlloc-m0.TotalAlloc)
fmt.Println("HeapAlloc: ", m1.HeapAlloc-m0.HeapAlloc)
```

  - debugging the code, I see `181407` categories that are queried and unmarshaled.
  - The results of a few runs are as follow:
  
| Time                  | Alloc          | TotalAlloc   | HeapAlloc
| --------------- | ---------- | ----------- | ------------
| 5.404870996s | 33181576 | 82444032  | 33181576
| 7.233135859s | 33182528 | 82444000  | 33182528
| 7.479702024s | 33183840 | 82444632  | 33183840
| 6.505728501s | 33183672 | 82443600  | 33183672
| 8.37185516s | 33185192 | 82444600  | 33185192
| 5.802228546s | 33186392 | 82443704  | 33186392
| 5.238508307s | 33185024 | 82444120  | 33185024
| 6.788198018s | 33185080 | 82443720  | 33185080
| 6.428873483s | 33182512 | 82444504  | 33182512
| 5.019777482s | 33182960 | 82441504  | 33182960

### Who can review

anyone
